app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,247 @@
+module ArcgisApi
+  # Struct to store token information, this allows us to track
+  # real expiration time with various rails cache backends many of them
+  # do not support entry expiration.
+  # @member token the token string
+  # @member expires_at hard expiration timestamp in epoch seconds
+  # @member sliding_expires_at optional the token keeper to maintain for
```

Comment:
@svalexander some description here https://brokul.dev/authentication-cookie-lifetime-and-sliding-expiration, not exactly with dotnet, but similar concept, basically avoiding a lot of clients getting token at the time very close to it's expiration time.

Reply:
The sliding expiration doesn't seem to fit the use case as well since we have a regularly scheduled job to refresh the token and don't have to worry about user actions to trigger a refresh?

Reply:
As I mentioned elsewhere, I think sliding expiration is a bit more complicated we require for this use case. Mostly we need to renew the token before it expires & before the IDP attempts to renew it. I'd be ok removing the frontend IDP renewal altogether if we can have a relatively high frequency & concurrency controls on the job.